File: ntics/ntics_registration.pas
unit ntics_registration;
interface
uses Windows, SysUtils, Forms, Classes;

procedure SetKey(dkey,nkey:string);
procedure SetSerial(FileName:string);
function GetFirmName:string; {return COMPANY_NAME;}
function GetEDRPOU:string; {return EDRPOU;}
function GetTFO:integer; {return ID_FIZ;}
function GetLicenze():integer; // return current licenze number
function GetLicenzeName():string; // return current licenze ASCII name
procedure AddLicenze(licenze:string); // add new licenze to licenze list



implementation
uses FGInt, FGIntRSA, ntics_const, ntics_storage;

const
  ID_FIZ:Integer = 0;
  LICENZE:Integer = 0;
  EDRPOU:string[11] = 'None';
  COMPANY_NAME:string[100] = 'None';
  SERIAL:string[100] = 'None';
  KeyExist:boolean = false;
  delimiter:string = ';';
  ShareWare = 'ShareWare';

var
  FD:TFGInt;
  FN:TFGInt;
  LicenzeName:array of string;



procedure Initialize();
var
  Null:TFGInt;
  s:string;
begin
  // Розшифровка ключа
  s := NetStorage.ReadString('OPTIONS','SERIAL','');
  if s <> '' then
  begin
    Base10StringToFGInt(s,Null);
    FGIntToBase256String(Null,s);
    FGIntDestroy(Null);
    RSADecrypt(s,FD,FN,Null,Null,Null,Null,s);

    // проверка версии ключа
    if  Pos(s,delimiter)=0 then Exit;
    if Copy(s,1,Pos(s,delimiter)-1) <> VersionKey then Exit;
    Delete(s,1,Pos(s,delimiter));

    if  Pos(s,delimiter)=0 then Exit;
    if Copy(s,1,Pos(s,delimiter)-1) <> Copyrigth then Exit;
    Delete(s,1,Pos(s,delimiter));

    // загрузка значений
    if  Pos(s,delimiter)=0 then Exit;
    EDRPOU := Copy(s,1,Pos(s,delimiter)-1);
    Delete(s,1,Pos(s,delimiter));

    if  Pos(s,delimiter)=0 then Exit;
    COMPANY_NAME := Copy(s,1,Pos(s,delimiter)-1);
    Delete(s,1,Pos(s,delimiter));

    if  Pos(s,delimiter)=0 then Exit;
    ID_FIZ := StrToInt(Copy(s,1,Pos(s,delimiter)-1));
    Delete(s,1,Pos(s,delimiter));

    if  Pos(s,delimiter)=0 then Exit;
    LICENZE := StrToInt(Copy(s,1,Pos(s,delimiter)-1));
    Delete(s,1,Pos(s,delimiter));

    if  Pos(s,delimiter)=0 then Exit;
    SERIAL := Copy(s,1,Pos(s,delimiter)-1);
  end;
end;

procedure SetKey(dkey,nkey:string);
begin
  Base10StringToFGInt(dkey,FD);
  Base10StringToFGInt(nkey,FN);
  Initialize();
end;

procedure SetSerial(FileName:string);
var
  f:Text;
  s:string;
begin
  AssignFile(f,FileName);
  Reset(f);
  Readln(f,s);
  if s<>'' then
    NetStorage.WriteString('OPTIONS','SERIAL',s);
  Initialize();
end;

function GetFirmName:string;
begin
  Result:=COMPANY_NAME;
end;

function GetEDRPOU:string;
begin
  Result:=EDRPOU;
end;

function GetTFO:integer;
begin
  Result:=ID_FIZ;
end;

// Работа с лицензиями
function GetLicenze():integer;
begin
  if LICENZE>(Length(LicenzeName)-1) then
    Result:=0
  else Result:=LICENZE
end;

function GetLicenzeName():string;
begin
  Result:=LicenzeName[GetLicenze()];
end;

procedure AddLicenze(licenze:string);
begin
  SetLength(LicenzeName,Length(LicenzeName)+1);
  LicenzeName[Length(LicenzeName)-1] := licenze;
end;


initialization
  SetLength(LicenzeName,1);
  LicenzeName[0]:=ShareWare;

end.

File: NTICSLib/source/ntics_registration.cpp
//---------------------------------------------------------------------------
#include <vcl.h>
#include "ntics_includes.h"
#pragma hdrstop

#include "ntics_registration.h"
#pragma package(smart_init)


// название основного конфигурационного файла
String GetDefaultIniName()
{
  String s = ExtractFileName(Application->ExeName);
  return ExtractFilePath(Application->ExeName) + s.SubString(1,s.Pos(".")) + ".cfg";
}

String Registration::StorageName = GetDefaultIniName();
String Registration::DefaultSection = "Defoult"; // Set function to return user name

String __fastcall Registration::LoadValue(String Section, String Ident, String Default)
{
  TIniFile* INIFile = new TIniFile(Registration::StorageName);
  String s = INIFile->ReadString(Section, Ident, "0"+Default);

  s.Delete(1,1);
  delete INIFile;
  return s;
}


String __fastcall Registration::LoadValue(String Ident, String Default)
{
 return Registration::LoadValue(Registration::DefaultSection, Ident, Default);
}

void __fastcall Registration::SaveValue(String Section, String Ident, String Value, bool Crypt)
{
  TIniFile* INIFile = new TIniFile(Registration::StorageName);
  if (Crypt)
  {
   Value = "1" + Value; // insert this coder
  }
  else Value = "0" + Value;
  INIFile->WriteString(Section, Ident, Value);
  delete INIFile;
}

void __fastcall Registration::EraseSection(String Section)
{
  TIniFile* INIFile = new TIniFile(Registration::StorageName);
  INIFile->EraseSection(Section);
  delete INIFile;
}

void __fastcall Registration::SaveValue(String Ident, String Value, bool Crypt)
{
  Registration::SaveValue(Registration::DefaultSection, Ident, Value, Crypt);
}


int ID_FIZ = 0;
unsigned int LICENZE = 0;
char EDRPOU[11] = "None";
char COMPANY_NAME[100] = "None";
char SERIAL[100] = "None";
Fgint::TFGInt FD;
Fgint::TFGInt FN;
bool KeyExist = false;


void __fastcall Initialize()
{
  // Розшифровка ключа
  if (Registration::LoadValue("OPTIONS","SERIAL","") != "")
  {
    TFGInt Null;
    String delimiter = ";";
    String s = Registration::LoadValue("OPTIONS","SERIAL","");
    Base10StringToFGInt(s,Null);
    FGIntToBase256String(Null,s);
    FGIntDestroy(Null);
    RSADecrypt(s,FD,FN,Null,Null,Null,Null,s);

    // проверка версии ключа
    if (!s.Pos(delimiter)) return;
    if (s.SubString(1,s.Pos(delimiter)-1) != VersionKey) return;
    s.Delete(1,s.Pos(delimiter));

    if (!s.Pos(delimiter)) return;
    if (s.SubString(1,s.Pos(delimiter)-1) != Copyrigth) return;
    s.Delete(1,s.Pos(delimiter));

    // загрузка значений
    if (!s.Pos(delimiter)) return;
    StrCopy(EDRPOU,s.SubString(1,s.Pos(delimiter)-1).c_str());
    s.Delete(1,s.Pos(delimiter));

    if (!s.Pos(delimiter)) return;
    StrCopy(COMPANY_NAME,s.SubString(1,s.Pos(delimiter)-1).c_str());
    s.Delete(1,s.Pos(delimiter));

    if (!s.Pos(delimiter)) return;
    ID_FIZ = StrToInt(s.SubString(1,s.Pos(delimiter)-1));
    s.Delete(1,s.Pos(delimiter));

    if (!s.Pos(delimiter)) return;
    LICENZE = StrToInt(s.SubString(1,s.Pos(delimiter)-1));
    s.Delete(1,s.Pos(delimiter));

    if (!s.Pos(delimiter)) return;
    StrCopy(SERIAL,s.SubString(1,s.Pos(delimiter)-1).c_str());
    //s.Delete(1,s.Pos(delimiter));
  }
}

void __fastcall Registration::SetKey(const char* dkey, const char* nkey)
{
  Base10StringToFGInt(dkey,FD);
  Base10StringToFGInt(nkey,FN);
  Initialize();
}

void __fastcall Registration::SetSerial(String FileName)
{
  ifstream f;
  f.open(FileName.c_str());
  if (f)
  {
    char buff[1000];
    f.getline(buff,sizeof(buff));
    Registration::SaveValue("OPTIONS","SERIAL",buff);
    Initialize();
  }
}


char* __fastcall Registration::GetFirmName(){return COMPANY_NAME;}
char* __fastcall Registration::GetEDRPOU(){return EDRPOU;}
int __fastcall Registration::GetTFO(){return ID_FIZ;}

// Работа с лицензиями
const char* ShareWare = "ShareWare";
vector<const char*> LicenzeName(1,ShareWare);

int __fastcall Registration::GetLicenze()
{return (LICENZE > LicenzeName.size())?0:LICENZE;}

const char* __fastcall Registration::GetLicenzeName()
{
  return LicenzeName[Registration::GetLicenze()];
}

void __fastcall Registration::AddLicenze(const char* licenze)
{
  LicenzeName.resize(LicenzeName.size()+1);
  LicenzeName[LicenzeName.size()-1] = licenze;
}